File: fixtures/scale/dart/large/lib/ciphers/feistel/feistel_network.dart
import 'dart:typed_data';

import '../../util/block_codec.dart';
import 'feistel_key.dart';

/// Balanced 16-round Feistel permutation over 8-byte blocks.
class FeistelNetwork {
  final FeistelKey key;

  FeistelNetwork(this.key);

  void block(Uint8List data, int off, bool reverse) {
    final packed = readBlock(data, off);
    var left = (packed >> 32) & 0xFFFFFFFF;
    var right = packed & 0xFFFFFFFF;
    for (var r = 0; r < 16; r++) {
      final round = reverse ? 15 - r : r;
      final tmp = right;
      right = left ^ _roundFn(right, key.subKey(round));
      left = tmp;
    }
    writeBlock(data, off, ((right & 0xFFFFFFFF) << 32) | (left & 0xFFFFFFFF));
  }

  int _roundFn(int half, int subKey) {
    final rotated = ((half ^ subKey) << 5 | (half ^ subKey) >> 27) & 0xFFFFFFFF;
    return (rotated * 0x1E3779B9 + subKey) & 0xFFFFFFFF;
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/feistel/FeistelNetwork.java
package cryptobench.ciphers.feistel;

import cryptobench.util.BlockCodec;

/** Balanced 16-round Feistel permutation over 8-byte blocks. */
public class FeistelNetwork {
    private final FeistelKey key;

    public FeistelNetwork(FeistelKey key) {
        this.key = key;
    }

    public void block(byte[] data, int off, boolean reverse) {
        long packed = BlockCodec.read(data, off);
        int left = (int) (packed >>> 32);
        int right = (int) packed;
        for (int r = 0; r < 16; r++) {
            int round = reverse ? 15 - r : r;
            int tmp = right;
            right = left ^ round(right, key.subKey(round));
            left = tmp;
        }
        BlockCodec.write(data, off, ((long) right << 32) | (left & 0xFFFFFFFFL));
    }

    private int round(int half, int subKey) {
        int mixed = Integer.rotateLeft(half ^ subKey, 5);
        return mixed * 0x9E3779B9 + subKey;
    }
}
